csrc/multidevice/utils.cpp

```diff
@@ -266,40 +235,65 @@ void insertReshardings(Fusion* fusion) {
         expr->outputs().size() == 1,
         "multi-output expressions are not supported");
     auto output = expr->outputs().at(0)->as<TensorView>();
-    std::vector<TensorView*> new_inputs;
-    for (auto input : getTvsWithDifferentSharding(
-             output, ir_utils::filterByType<TensorView>(expr->inputs()))) {
-      // TODO: reuse cacheAfter?
-      // TODO: here we should add a mechanism to potentially reuse the inserted
-      // resharding accross all the consumer of the resharded tensor. This way
-      // we could avoid wasteful resharding set insertion.
-      TensorView* new_input = set(input);
-      new_inputs.push_back(new_input);
-      expr = ir_utils::replaceValInExprInputs(expr, input, new_input);
+
+    auto inputs = getTvsWithDifferentSharding(
+        output, ir_utils::filterByType<TensorView>(expr->inputs()));
+
+    // Insert resharding set after the expr when there is only one input.
```

Comment:
Why is this better than inserting the resharding before? It sounds like we are lacking some heuristics to look at tensor sizes?

Reply:
Yes, we need to add better heuristics of where to insert the sets. Same for your below comment. Right now we don't look at minimizing communication size or what the effect is on the surrounding computation.

Reply:
Got it. Can you add a code comment to say heuristics need to be retuned in the future? While no heuristics are perfect, the decisions as they are seem a bit too arbitrary and can cause confusion without a code comment.